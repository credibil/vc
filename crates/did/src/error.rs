/// DID resolution error codes
#[derive(thiserror::Error, Debug)]
pub enum Error {
    /// The DID method is not supported.
    #[error("methodNotSupported")]
    MethodNotSupported(String),

    /// The DID supplied to the DID resolution function does not conform to
    /// valid syntax.
    #[error("invalidDid")]
    InvalidDid(String),

    /// The DID resolver was unable to find the DID document resulting from
    /// this resolution request.
    #[error("notFound")]
    NotFound(String),

    /// The representation requested via the accept input metadata property is
    /// not supported by the DID method and/or DID resolver.
    #[error("representationNotSupported")]
    RepresentationNotSupported(String),

    /// The DID URL is invalid
    #[error("invalidDidUrl")]
    InvalidDidUrl(String),

    // ---- Creation Errors ----  //
    /// The byte length of raw public key does not match that expected for the
    /// associated multicodecValue.
    #[error("invalidPublicKeyLength")]
    InvalidPublicKeyLength(String),

    /// The public key is invalid
    #[error("invalidPublicKey")]
    InvalidPublicKey(String),

    /// Public key format is not known to the implementation.
    #[error("unsupportedPublicKeyType")]
    UnsupportedPublicKeyType(String),

    /// Other, unspecified errors.
    #[error(transparent)]
    Other(#[from] anyhow::Error),
}

impl Error {
    /// Returns the error code.
    #[must_use]
    pub fn code(&self) -> String {
        self.to_string()
    }

    /// Returns the associated error message.
    #[must_use]
    pub fn message(&self) -> String {
        match self {
            Self::MethodNotSupported(msg)
            | Self::InvalidDid(msg)
            | Self::NotFound(msg)
            | Self::InvalidDidUrl(msg)
            | Self::RepresentationNotSupported(msg)
            | Self::InvalidPublicKeyLength(msg)
            | Self::InvalidPublicKey(msg)
            | Self::UnsupportedPublicKeyType(msg) => msg.clone(),
            Self::Other(err) => err.to_string(),
        }
    }
}

// impl From<anyhow::Error> for Error {
//     fn from(err: anyhow::Error) -> Self {
//         Self
//     }
// }

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn error_code() {
        let err = Error::MethodNotSupported("Method not supported".into());
        assert_eq!(err.message(), "Method not supported");
    }
}
