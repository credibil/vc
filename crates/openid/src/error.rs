//! # `OpenID` Errors
//!
//! This module defines errors for `OpenID` for Verifiable Credential Issuance
//! and Verifiable Presentations.

// TODO: add support for "client-state" in error responses.
// TODO: use customer serialisation for Err enum.

use std::fmt::Debug;

// use anyhow::Error;
use serde::{Deserialize, Serialize, Serializer};
use thiserror::Error;

/// `OpenID` error codes for  for Verifiable Credential Issuance and Presentation.
#[derive(Error, Debug, Deserialize)]
pub enum Error {
    /// The request is missing a required parameter, includes an unsupported
    /// parameter value, repeats a parameter, includes multiple credentials,
    /// utilizes more than one mechanism for authenticating the client, or is
    /// otherwise malformed.
    #[error(r#"{{"error": "invalid_request", "error_description": "{0}"}}"#)]
    InvalidRequest(String),

    /// Client authentication failed (e.g., unknown client, no client
    /// authentication included, or unsupported authentication method).
    ///
    /// The client tried to send a Token Request with a Pre-Authorized Code
    /// without Client ID but the Authorization Server does not support
    /// anonymous access.
    ///
    /// For Verifiable Presentations:
    ///
    /// `client_metadata` or `client_metadata_uri` is set, but the Wallet
    /// recognizes Client Identifier and already knows metadata associated
    /// with it.
    ///
    /// Verifier's pre-registered metadata has been found based on the Client
    /// Identifier, but `client_metadata` parameter is also set.
    #[error(r#"{{"error": "invalid_client", "error_description": "{0}"}}"#)]
    InvalidClient(String),

    /// The provided authorization grant (e.g., authorization code,
    /// pre-authorized_code) or refresh token is invalid, expired, revoked,
    /// does not match the redirection URI used in the authorization
    /// request, or was issued to another client.
    ///
    /// The Authorization Server expects a PIN in the pre-authorized flow but
    /// the client provides the wrong PIN.
    #[error(r#"{{"error": "invalid_grant", "error_description": "{0}"}}"#)]
    InvalidGrant(String),

    /// The client is not authorized to request an authorization code using this
    /// method.
    #[error(r#"{{"error": "unauthorized_client", "error_description": "{0}"}}"#)]
    UnauthorizedClient(String),

    /// The authorization grant type is not supported by the authorization
    /// server.
    #[error(r#"{{"error": "unsupported_grant_type", "error_description": "{0}"}}"#)]
    UnsupportedGrantType(String),

    /// The requested scope is invalid, unknown, malformed, or exceeds the scope
    /// granted.
    #[error(r#"{{"error": "invalid_scope", "error_description": "{0}"}}"#)]
    InvalidScope(String),

    /// The resource owner or authorization server denied the request.
    #[error(r#"{{"error": "access_denied", "error_description": "{0}"}}"#)]
    AccessDenied(String),

    /// The authorization server does not support obtaining an authorization
    /// code using this method.
    #[error(r#"{{"error": "unsupported_response_type", "error_description": "{0}"}}"#)]
    UnsupportedResponseType(String),

    /// The authorization server encountered an unexpected condition that
    /// prevented it from fulfilling the request.
    #[error(r#"{{"error": "server_error", "error_description": "{0}"}}"#)]
    ServerError(String),

    /// The authorization server is unable to handle the request due to
    /// temporary overloading or maintenance.
    #[error(r#"{{"error": "temporarily_unavailable", "error_description": "{0}"}}"#)]
    TemporarilyUnavailable(String),

    /// ------------------------------
    /// Verifiable Credential Issuance
    /// ------------------------------

    /// Credential Endpoint:

    /// The Credential Request is missing a required parameter, includes an unsupported
    /// parameter or parameter value, repeats the same parameter, or is otherwise
    /// malformed.
    #[error(r#"{{"error": "invalid_credential_request", "error_description": "{0}"}}"#)]
    InvalidCredentialRequest(String),

    /// Requested credential type is not supported.
    #[error(r#"{{"error": "unsupported_credential_type", "error_description": "{0}"}}"#)]
    UnsupportedCredentialType(String),

    /// Requested credential format is not supported.
    #[error(r#"{{"error": "unsupported_credential_format", "error_description": "{0}"}}"#)]
    UnsupportedFormat(String),

    /// Credential Request did not contain a proof, or proof was invalid, i.e. it was
    /// not bound to a Credential Issuer provided `c_nonce`. The error response contains
    /// new `c_nonce` as well as `c_nonce_expires_in` values to be used by the Wallet
    /// when creating another proof of possession of key material.
    #[allow(missing_docs)]
    #[error(r#"{{"error": "invalid_proof", "error_description": "{hint}", "c_nonce": "{c_nonce}", "c_nonce_expires_in": {c_nonce_expires_in}}}"#)]
    InvalidProof { hint: String, c_nonce: String, c_nonce_expires_in: i64 },

    /// This error occurs when the encryption parameters in the Credential Request are
    /// either invalid or missing. In the latter case, it indicates that the Credential
    /// Issuer requires the Credential Response to be sent encrypted, but the Credential
    /// Request does not contain the necessary encryption parameters.
    #[error(r#"{{"error": "invalid_encryption_parameters", "error_description": "{0}"}}"#)]
    InvalidEncryptionParameters(String),

    /// Deferred Issuance Endpoint:

    /// The Credential issuance is still pending. The error response SHOULD also contain
    /// the interval member, determining the minimum amount of time in seconds that the
    /// Wallet needs to wait before providing a new request to the Deferred Credential
    /// Endpoint. If interval member is missing or its value is not provided, the Wallet
    /// MUST use 5 as the default value.
    #[error(r#"{{"error": "issuance_pending", "error_description": "{0}"}}"#)]
    IssuancePending(String),

    /// The Deferred Credential Request contains an invalid `transaction_id`. This error
    /// occurs when the `transaction_id` was not issued by the respective Credential
    /// Issuer or it was already used to obtain the Credential.
    #[error(r#"{{"error": "invalid_transaction_id", "error_description": "{0}"}}"#)]
    InvalidTransactionId(String),

    /// ------------------------------
    /// Verifiable Presentation
    /// ------------------------------

    /// The Wallet does not support any of the formats requested by the
    /// Verifier, such as those included in the `vp_formats` registration
    /// parameter.
    #[error(r#"{{"error": "vp_formats_not_supported", "error_description": "{0}"}}"#)]
    VpFormatsNotSupported(String),

    /// The Presentation Definition URL cannot be reached.
    #[error(r#"{{"error": "invalid_presentation_definition_uri", "error_description": "{0}"}}"#)]
    InvalidPresentationDefinitionUri(String),

    /// The Presentation Definition URL can be reached, but the specified
    /// `presentation_definition` cannot be found at the URL.
    #[error(
        r#"{{"error": "invalid_presentation_definition_reference", "error_description": "{0}"}}"#
    )]
    InvalidPresentationDefinitionReference(String),

    /// The Wallet appears to be unavailable and therefore unable to respond to the
    /// request.
    /// Use when the User Agent cannot invoke the Wallet and another component
    /// receives the request while the End-User wishes to continue the journey on
    /// the Verifier website.
    #[error(r#"{{"error": "wallet_unavailable", "error_description": "{0}"}}"#)]
    WalletUnavailable(String),
}

/// Error response for `OpenID` for Verifiable Credentials.
#[allow(clippy::module_name_repetitions)]
#[derive(Deserialize, Serialize)]
pub struct OidError {
    /// Error code.
    pub error: String,

    /// Error description.
    pub error_description: String,

    /// Optional client-state parameter.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub state: Option<String>,

    /// A fresh `c_nonce` to use when retrying Proof submission.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub c_nonce: Option<String>,

    /// The expiry time of the `c_nonce`.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub c_nonce_expires_in: Option<i64>,
}

impl Serialize for Error {
    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
        use serde::ser::Error as SerdeError;

        let Ok(error) = serde_json::from_str::<OidError>(&self.to_string()) else {
            return Err(SerdeError::custom("issue deserializing Err"));
        };
        error.serialize(serializer)
    }
}

impl Error {
    /// Transfrom error to `OpenID` compatible json format.
    #[must_use]
    pub fn to_json(self) -> serde_json::Value {
        serde_json::from_str(&self.to_string()).unwrap_or_default()
    }

    /// Transfrom error to `OpenID` compatible query string format.
    /// Does not include `c_nonce` as this is not required for in query
    /// string responses.
    #[must_use]
    pub fn to_querystring(self) -> String {
        serde_qs::to_string(&self).unwrap_or_default()
    }
}

#[cfg(test)]
mod test {
    use serde_json::{json, Value};

    use super::*;

    // Test that error details are retuned as json.
    #[test]
    fn err_json() {
        let err = Error::InvalidRequest("bad request".into());
        let ser: Value = serde_json::from_str(&err.to_string()).unwrap();
        assert_eq!(ser, json!({"error":"invalid_request", "error_description": "bad request"}));
    }

    // Test that the error details are returned as an http query string.
    #[test]
    fn err_querystring() {
        let err = Error::InvalidRequest("Invalid request description".into());
        let ser = serde_qs::to_string(&err).unwrap();
        assert_eq!(ser, "error=invalid_request&error_description=Invalid+request+description");
    }

    // Test that the error details are returned as an http query string.
    #[test]
    fn err_serialize() {
        let err = Error::InvalidRequest("bad request".into());
        let ser = serde_json::to_value(&err).unwrap();
        assert_eq!(ser, json!({"error":"invalid_request", "error_description": "bad request"}));
    }

    // Test an InvalidProof error returns c_nonce and c_nonce_expires_in values
    // in the external response.
    #[test]
    fn proof_err() {
        let err = Error::InvalidProof {
            hint: "".into(),
            c_nonce: "c_nonce".into(),
            c_nonce_expires_in: 10,
        };
        let ser: Value = serde_json::from_str(&err.to_string()).unwrap();

        assert_eq!(
            ser,
            json!({
                "error": "invalid_proof",
                "error_description": "",
                "c_nonce": "c_nonce",
                "c_nonce_expires_in": 10,
            })
        );
    }
}
