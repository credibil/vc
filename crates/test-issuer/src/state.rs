#![allow(missing_docs)]

use std::collections::HashMap;
use std::sync::{Arc, Mutex};

use anyhow::anyhow;
use chrono::{DateTime, Utc};
use credibil_vc::openid::provider::Result;
use serde::Serialize;
use serde::de::DeserializeOwned;

#[derive(Default, Clone, Debug)]
pub struct Store {
    store: Arc<Mutex<HashMap<String, Vec<u8>>>>,
}

impl Store {
    #[must_use]
    pub fn new() -> Self {
        Self {
            store: Arc::new(Mutex::new(HashMap::new())),
        }
    }

    #[allow(clippy::unnecessary_wraps)]
    pub fn put(&self, key: &str, state: impl Serialize, _: DateTime<Utc>) -> Result<()> {
        let state = serde_json::to_vec(&state)?;
        self.store.lock().expect("should lock").insert(key.to_string(), state);
        Ok(())
    }

    pub fn get<T: DeserializeOwned>(&self, key: &str) -> Result<T> {
        let Some(state) = self.store.lock().expect("should lock").get(key).cloned() else {
            return Err(anyhow!("state not found for key: {key}"));
        };
        Ok(serde_json::from_slice(&state)?)
    }

    #[allow(clippy::unnecessary_wraps)]
    pub fn purge(&self, key: &str) -> Result<()> {
        self.store.lock().expect("should lock").remove(key);
        Ok(())
    }
}
