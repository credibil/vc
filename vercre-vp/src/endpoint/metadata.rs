//! # Metadata Endpoint
//!
//! This endpoint is used to make Verifier metadata available to the Wallet.
//!
//! As the Verifier is a client to the Wallet's Authorization Server, this endpoint
//! returns Client metadata as defined in [RFC7591].
//!
//! [RFC7591]: https://www.rfc-editor.org/rfc/rfc7591.html

use std::fmt::Debug;

use tracing::{instrument, trace};
use vercre_core::vp::{MetadataRequest, MetadataResponse};
use vercre_core::{Callback, Client, Result, Signer, StateManager};

use super::Endpoint;

/// Metadata request handler.
impl<P> Endpoint<P>
where
    P: Client + StateManager + Signer + Callback + Clone + Debug,
{
    /// Endpoint for Wallets to request Verifier (Client) metadata.
    ///
    /// # Errors
    ///
    /// Returns an `OpenID4VP` error if the request is invalid or if the provider is
    /// not available.
    pub async fn metadata(&self, request: impl Into<MetadataRequest>) -> Result<MetadataResponse> {
        self.handle_request(request.into(), Context {}).await
    }
}

#[derive(Debug)]
struct Context;

impl super::Context for Context {
    type Request = MetadataRequest;
    type Response = MetadataResponse;

    // TODO: return callback_id
    fn callback_id(&self) -> Option<String> {
        None
    }

    #[instrument]
    async fn process<P>(&self, provider: &P, req: &Self::Request) -> Result<Self::Response>
    where
        P: Client + Debug,
    {
        trace!("Context::process");

        Ok(MetadataResponse {
            client: Client::metadata(provider, &req.client_id).await?,
        })
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_yaml_snapshot as assert_snapshot;
    use test_utils::vp_provider::Provider;

    // use test_utils::wallet::CLIENT_ID;
    use super::*;

    #[tokio::test]
    async fn metadata_ok() {
        test_utils::init_tracer();
        let provider = Provider::new();

        let request = MetadataRequest {
            client_id: "http://credibil.io".to_string(),
        };
        let response = Endpoint::new(provider).metadata(request).await.expect("response is ok");
        assert_snapshot!("response", response, {
            ".vp_formats" => insta::sorted_redaction()
        });
    }
}
