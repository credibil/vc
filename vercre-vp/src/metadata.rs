//! # Metadata Endpoint
//!
//! This endpoint is used to make Verifier metadata available to the Wallet.
//!
//! As the Verifier is a client to the Wallet's Authorization Server, this endpoint
//! returns Client metadata as defined in [RFC7591](https://www.rfc-editor.org/rfc/rfc7591).

use std::fmt::Debug;

use tracing::instrument;
pub use vercre_core::metadata as types;
use vercre_core::provider::{Callback, Client, Signer, StateManager};
#[allow(clippy::module_name_repetitions)]
pub use vercre_core::vp::{MetadataRequest, MetadataResponse};
use vercre_core::Result;

use super::Endpoint;

/// Metadata request handler.
impl<P> Endpoint<P>
where
    P: Client + StateManager + Signer + Callback + Clone + Debug,
{
    /// Endpoint for Wallets to request Verifier (Client) metadata.
    ///
    /// # Errors
    ///
    /// Returns an `OpenID4VP` error if the request is invalid or if the provider is
    /// not available.
    #[instrument(level = "debug", skip(self))]
    pub async fn metadata(&self, request: &MetadataRequest) -> Result<MetadataResponse> {
        let ctx = Context {
            _p: std::marker::PhantomData,
        };

        vercre_core::Endpoint::handle_request(self, request, ctx).await
    }
}

#[derive(Debug)]
struct Context<P> {
    _p: std::marker::PhantomData<P>,
}

impl<P> vercre_core::Context for Context<P>
where
    P: Client + StateManager + Signer + Callback + Clone + Debug,
{
    type Provider = P;
    type Request = MetadataRequest;
    type Response = MetadataResponse;

    // TODO: return callback_id
    fn callback_id(&self) -> Option<String> {
        None
    }

    async fn process(&self, provider: &P, req: &Self::Request) -> Result<Self::Response> {
        tracing::debug!("Context::process");

        Ok(MetadataResponse {
            client: Client::metadata(provider, &req.client_id).await?,
        })
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_yaml_snapshot as assert_snapshot;
    use test_utils::vp_provider::Provider;

    // use test_utils::wallet_provider::wallet::CLIENT_ID;
    use super::*;

    #[tokio::test]
    async fn metadata_ok() {
        test_utils::init_tracer();
        let provider = Provider::new();

        let request = MetadataRequest {
            client_id: String::from("http://vercre.io"),
        };
        let response = Endpoint::new(provider).metadata(&request).await.expect("response is ok");
        assert_snapshot!("response", response, {
            ".vp_formats" => insta::sorted_redaction()
        });
    }
}
