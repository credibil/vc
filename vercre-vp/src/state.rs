//! State is used by the library to persist request information between steps
//! in the issuance process.
use chrono::{DateTime, Duration, Utc};
use derive_builder::Builder;
use serde::{Deserialize, Serialize};
use vercre_core::error::Err;
use vercre_core::vp::RequestObject;
use vercre_core::{err, Result};

pub(crate) enum Expire {
    Request,
    // Nonce,
}

impl Expire {
    pub(crate) fn duration(&self) -> Duration {
        match self {
            Expire::Request => Duration::minutes(5),
        }
    }
}

#[derive(Builder, Clone, Debug, Default, Deserialize, Serialize)]
pub(crate) struct State {
    /// The time this state item should expire.
    #[builder(default = "Utc::now() + Expire::Request.duration()")]
    pub(crate) expires_at: DateTime<Utc>,

    /// The Verifier's Request Object. Saved for use by the 'request_uri' endpoint
    /// and in comparing the Presentation Definition to the Presentation Submission.
    pub(crate) request_object: RequestObject,

    // The callback ID for the current request.
    #[serde(skip_serializing_if = "Option::is_none")]
    #[builder(default)]
    pub(crate) callback_id: Option<String>,
}

impl State {
    /// Returns a new [`StateBuilder`], which can be used to build a [State]
    #[must_use]
    pub(crate) fn builder() -> StateBuilder {
        StateBuilder::default()
    }

    /// Serializes this [`State`] object into a byte array.
    pub(crate) fn to_vec(&self) -> Vec<u8> {
        // TODO: return Result instead of panicking
        match serde_json::to_vec(self) {
            Ok(res) => res,
            Err(e) => panic!("Failed to serialize state: {e}"),
        }
    }

    pub(crate) fn from_slice(value: &[u8]) -> Result<Self> {
        match serde_json::from_slice::<Self>(value) {
            Ok(res) => {
                if res.expired() {
                    err!(Err::InvalidRequest, "State has expired");
                }
                Ok(res)
            }
            Err(e) => err!(Err::ServerError(e.into()), "Failed to deserialize state"),
        }
    }

    /// Determines whether state has expired or not.
    pub(crate) fn expired(&self) -> bool {
        self.expires_at.signed_duration_since(Utc::now()).num_seconds() < 0
    }
}

impl TryFrom<&[u8]> for State {
    type Error = vercre_core::error::Error;

    fn try_from(value: &[u8]) -> Result<Self> {
        State::from_slice(value)
    }
}

impl TryFrom<Vec<u8>> for State {
    type Error = vercre_core::error::Error;

    fn try_from(value: Vec<u8>) -> Result<Self> {
        State::try_from(value.as_slice())
    }
}

// /// [`StateBuilder`] is used to build a [State]
// #[derive(Clone, Default)]
// #[allow(clippy::module_name_repetitions)]
// pub struct StateBuilder {
//     pub(crate) state: State,
// }

// /// [`StateBuilder`] is used to build a [`State`] item.
// #[allow(dead_code)]
// impl StateBuilder {
//     /// Returns a new [`StateBuilder`]
//     #[must_use]
//     pub(crate) fn new() -> Self {
//         Self {
//             state: State {
//                 expires_at: Utc::now() + Expire::Request.duration(),
//                 ..Default::default()
//             },
//         }
//     }

//     /// Sets the `expires_at` property
//     #[must_use]
//     pub(crate) fn expires_at(mut self, expires_at: DateTime<Utc>) -> Self {
//         self.state.expires_at = expires_at;
//         self
//     }

//     /// Sets the `request_object` property
//     #[must_use]
//     pub(crate) fn request_object(mut self, obj: &RequestObject) -> Self {
//         self.state.request_object = obj.clone();
//         self
//     }

//     /// Turns this builder into a [`State`]
//     pub(crate) fn build(self) -> State {
//         self.state
//     }
// }
