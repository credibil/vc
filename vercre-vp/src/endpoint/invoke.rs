//! # Initiate Endpoint
//!
//! This endpoint is used to **prepare** an [RFC6749] Authorization Request to use
//! to request Verifiable Presentations from an End-User's Wallet.
//!
//! While based on the [OpenID4VP] specification, the endpoint only implements a subset
//! of the specification requirements as recommended in the [JWT VC Presentation
//! Profile]. Aside from reducing complexity, the profile enables greater presentation
//! interoperability between Wallets and Verifiers.
//!
//! The Verifier requests an Authorization Request be prepared by articulating the
//! Credential(s) desired using a Credential Definition and, optionally, specifying
//! the device flow that will be used.
//!
//! # Example
//!
//! ```json
//! {
//!     "credentials": [
//!         {"type": ["VerifiableCredential", "EmployeeIDCredential"]},
//!         {"type": ["VerifiableCredential", "CitizenshipCredential"]}
//!    ],
//!     "device_flow": "CrossDevice"
//! }
//! ```
//!
//! The prepared Authorization Request Object may be sent by value or by reference as
//! defined in JWT-Secured Authorization Request (JAR) [RFC9101]. If sent by value,
//! the Request Object is sent directly to the Wallet as a URL fragment in the
//! Authorization Request. If by reference, the Authorization Request will contain a
//! `request_uri` pointing to the prepared Request Object.
//!
//! [OpenID4VP]: https://openid.net/specs/openid-connect-verifiable-presentations-1_0.html
//! [JWT VC Presentation Profile]: https://identity.foundation/jwt-vc-presentation-profile
//! [RFC6749]: https://www.rfc-editor.org/rfc/rfc6749.html
//! [RFC9101]: https://www.rfc-editor.org/rfc/rfc9101.html

use std::collections::HashMap;
use std::convert::From;
use std::fmt::Debug;

// use serde_json::map::Map;
// use serde_json::Value;
use tracing::{instrument, trace};
use uuid::Uuid;
use vercre_core::error::Err;
use vercre_core::metadata::CredentialDefinition;
use vercre_core::vp::{DeviceFlow, InvokeRequest, InvokeResponse, RequestObject};
use vercre_core::w3c::vp::{
    Constraints, Field, Filter, FilterValue, Format, InputDescriptor, PresentationDefinition,
};
use vercre_core::{err, gen, Algorithm, Callback, Client, Result, Signer, StateManager};

use super::Endpoint;
use crate::state::State;

// TODO: request supported Client Identifier schemes from the Wallet
// TODO: add support for other Client Identifier schemes

// StaticConfigurationValues
// {
//   "authorization_endpoint": "openid4vp:",
//   "response_types_supported": [
//     "vp_token"
//   ],
//   "vp_formats_supported": {
//     "jwt_vp_json": {
//       "alg_values_supported": ["ES256"]
//     },
//     "jwt_vc_json": {
//       "alg_values_supported": ["ES256"]
//     }
//   },
//   "request_object_signing_alg_values_supported": [
//     "ES256"
//   ]
// }

/// Initiate Request handler.
impl<P> Endpoint<P>
where
    P: Client + StateManager + Signer + Callback + Clone + Debug,
{
    /// Initiate an Authorization Request flow.
    ///
    /// # Errors
    ///
    /// Returns an `OpenID4VP` error if the request is invalid or if the provider is
    /// not available.
    pub async fn initiate(&self, request: impl Into<InvokeRequest>) -> Result<InvokeResponse> {
        let request = request.into();
        let ctx = Context {
            callback_id: request.callback_id.clone(),
        };

        self.handle_request(request, ctx).await
    }
}

#[derive(Debug, Default)]
struct Context {
    callback_id: Option<String>,
}

impl super::Context for Context {
    type Request = InvokeRequest;
    type Response = InvokeResponse;

    fn callback_id(&self) -> Option<String> {
        self.callback_id.clone()
    }

    #[instrument]
    async fn verify<P>(&self, provider: &P, request: &Self::Request) -> Result<&Self>
    where
        P: Client + StateManager + Debug,
    {
        trace!("Context::verify");

        if request.credentials.is_empty() {
            err!(Err::InvalidRequest, "No credentials specified");
        }
        Ok(self)
    }

    #[instrument]
    async fn process<P>(&self, provider: &P, request: &Self::Request) -> Result<Self::Response>
    where
        P: Client + StateManager + Debug,
    {
        trace!("Context::process");

        // generate presentation_definition
        let def = self.build_def(&request.credentials);
        let state_key = gen::state_key();

        // get client metadata
        let Ok(client_meta) = Client::metadata(provider, &request.client_id).await else {
            err!(Err::InvalidRequest, "Invalid client_id");
        };

        let mut req_obj = RequestObject {
            response_type: "vp_token".to_string(),
            state: Some(state_key.clone()),
            nonce: gen::nonce(),
            presentation_definition: Some(def),
            client_metadata: Some(client_meta),
            client_id_scheme: Some("redirect_uri".to_string()),
            ..Default::default()
        };

        let mut response = InvokeResponse::default();

        // Response Mode "direct_post" is RECOMMENDED for cross-device flows.
        // TODO: replace hard-coded endpoints with Provider-set values
        if request.device_flow == DeviceFlow::CrossDevice {
            req_obj.response_mode = Some("direct_post".to_string());
            req_obj.client_id = format!("{}/post", request.client_id);
            req_obj.response_uri = Some(format!("{}/post", request.client_id));
            response.request_uri = Some(format!("{}/request/{state_key}", request.client_id));
        } else {
            req_obj.client_id = format!("{}/callback", request.client_id);
            response.request_object = Some(req_obj.clone());
        }

        // save request object in state
        let state = State::builder().request_object(&req_obj).build();
        StateManager::put(provider, &state_key, state.to_vec(), state.expires_at).await?;

        Ok(response)
    }
}

impl Context {
    #[instrument]
    fn build_def(&self, cred_defs: &[CredentialDefinition]) -> PresentationDefinition {
        trace!("creating presentation definition");

        let mut input_descs = Vec::<InputDescriptor>::new();

        for cred_def in cred_defs {
            let fields = vec![
                (Field {
                    // TODO: build JSONPath query properly
                    path: vec!["$.type".to_string()],
                    filter: Some(Filter {
                        type_: "string".to_string(),
                        value: FilterValue::Const("EmployeeIDCredential".to_string()),
                    }),

                    ..Field::default()
                }),
            ];

            input_descs.push(InputDescriptor {
                // TODO: check types == [VerifiableCredential, <specific type>]
                id: cred_def.type_[1].clone(),
                constraints: Constraints {
                    fields: Some(fields),
                    limit_disclosure: None,
                },
                name: None,
                purpose: None,
                format: None, // Some(HashMap::<String, Format>::new()),
            });
        }

        let fmt = Format {
            alg: Some(vec![Algorithm::ES256K.to_string()]),
            proof_type: None,
        };

        // presentation definition
        PresentationDefinition {
            id: Uuid::new_v4().to_string(),
            input_descriptors: input_descs,
            name: None,
            purpose: None,
            format: Some(HashMap::from([("jwt_vc".to_string(), fmt)])),
        }
    }
}

#[cfg(test)]
mod tests {
    use assert_let_bind::assert_let;
    use insta::assert_yaml_snapshot as assert_snapshot;
    use serde_json::json;
    use test_utils::vp_provider::Provider;

    use super::*;

    #[tokio::test]
    async fn same_device() {
        test_utils::init_tracer();
        let provider = Provider::new();

        // create offer to 'send' to the app
        let body = json!({
            "credentials": [
                {"type": ["VerifiableCredential", "EmployeeIDCredential"]},
                {"type": ["VerifiableCredential", "CitizenshipCredential"]}
            ],
            "device_flow": "SameDevice"
        });

        let mut request =
            serde_json::from_value::<InvokeRequest>(body).expect("should deserialize");
        request.client_id = "http://credibil.io".to_string();

        let response =
            Endpoint::new(provider.clone()).initiate(request).await.expect("response is ok");

        assert_eq!(response.request_uri, None);
        assert_let!(Some(req_obj), &response.request_object);

        // check redacted fields are present
        assert!(req_obj.client_metadata.is_some());
        assert!(req_obj.presentation_definition.is_some());

        // compare response with saved state
        let state_key = req_obj.state.as_ref().expect("has state");
        let buf = StateManager::get(&provider, state_key).await.expect("state exists");
        let state = State::try_from(buf).expect("state is valid");

        assert_eq!(req_obj.nonce, state.request_object.nonce);
        assert_snapshot!("sd-response", response, {
            ".request_object.presentation_definition"  => "[presentation_definition]",
            ".request_object.client_metadata" => "[client_metadata]",
            ".request_object.state" => "[state]",
            ".request_object.nonce" => "[nonce]",
        });
    }

    #[tokio::test]
    async fn cross_device() {
        test_utils::init_tracer();
        let provider = Provider::new();

        // create offer to 'send' to the app
        let body = json!({
            "credentials": [
                {"type": ["VerifiableCredential", "EmployeeIDCredential"]},
                {"type": ["VerifiableCredential", "CitizenshipCredential"]}
            ],
        });

        let mut request =
            serde_json::from_value::<InvokeRequest>(body).expect("should deserialize");
        request.client_id = "http://credibil.io".to_string();

        let response =
            Endpoint::new(provider.clone()).initiate(request).await.expect("response is ok");

        assert!(response.request_object.is_none());
        assert_let!(Some(req_uri), response.request_uri);

        // check state for RequestObject
        let state_key = req_uri.split('/').last().expect("has state");
        let buf = StateManager::get(&provider, state_key).await.expect("state exists");
        let state = State::try_from(buf).expect("state is valid");
        assert_snapshot!("cd-state", state, {
            ".expires_at" => "[expires_at]",
            ".request_object.presentation_definition"  => "[presentation_definition]",
            ".request_object.client_metadata" => "[client_metadata]",
            ".request_object.state" => "[state]",
            ".request_object.nonce" => "[nonce]",
        });
    }
}
