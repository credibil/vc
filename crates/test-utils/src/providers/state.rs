use std::collections::HashMap;
use std::sync::{Arc, Mutex};

use anyhow::anyhow;
use chrono::{DateTime, Utc};
use endpoint::Result;

#[derive(Default, Clone, Debug)]
pub struct Store {
    store: Arc<Mutex<HashMap<String, Vec<u8>>>>,
}

impl Store {
    pub fn new() -> Self {
        Self {
            store: Arc::new(Mutex::new(HashMap::new())),
        }
    }

    #[allow(clippy::unnecessary_wraps)]
    pub fn put(&self, key: &str, state: Vec<u8>, _: DateTime<Utc>) -> Result<()> {
        self.store.lock().expect("should lock").insert(key.to_string(), state);
        Ok(())
    }

    pub fn get(&self, key: &str) -> Result<Vec<u8>> {
        let Some(state) = self.store.lock().expect("should lock").get(key).cloned() else {
            return Err(anyhow!("state not found for key: {key}"));
        };
        Ok(state)
    }

    #[allow(clippy::unnecessary_wraps)]
    pub fn purge(&self, key: &str) -> Result<()> {
        self.store.lock().expect("should lock").remove(key);
        Ok(())
    }
}
