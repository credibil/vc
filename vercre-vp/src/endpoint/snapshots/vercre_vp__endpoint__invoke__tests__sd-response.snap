---
source: vercre-vp/src/endpoint/invoke.rs
assertion_line: 267
expression: response
---
request_object:
  response_type: vp_token
  client_id: "http://credibil.io/callback"
  nonce: "[nonce]"
  state: "[state]"
  presentation_definition: "[presentation_definition]"
  client_id_scheme: redirect_uri
  client_metadata: "[client_metadata]"

